### PR TITLE
tedge-apt-plugin: support settings to control if configuration files are kept or replaced when included in a debian package

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This PR proposes a solution to keep/overwrite the config files while installing the software packages using `tedge-apt` plugin.
The option to keep the old config file can be set using the `tedge` command as below:
```
tedge config set apt.dpk.options.config keepold
```
To `overwrite` the config files with new config files, one can use the below command:

```
tedge config set apt.dpk.options.config keepnew
```

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/2418

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (reubenmiller) [inline tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot:26]: Clean up commands should be run in a Teardown block because these steps won't be run if your previous assertion (e.g. `Should Contain`) fails, which might lead to unexpected issues in other tests since you are reusing the same device in both tests.

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (reubenmiller) [inline tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot:50]: You don't need this as you're interfacing with the `apt` sm-plugin directly and not using the cloud.

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (reubenmiller) [inline tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot:33]: I'm actually surprised calling the `apt` sm plugin works like this.

I think we should be calling it like how it would be done in a real situation, e.g.

```
/etc/tedge/sm-plugins/apt install sampledeb --file /setup/sampledeb_1.0.0_all.deb
```

**Review by reubenmiller — CHANGES_REQUESTED** [review]
Can you please add the source files how you generated the sampledeb package (e.g. nfpm config file and the script used to build them) to the test folder. Just place them next to where the .deb files are located.

This will make it easier to make adjustments to the package in the future.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs
Comment (reubenmiller) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:738]: We should add to the note how these settings are mapped to the apt command, as people familiar with apt can easy see what the tedge option is mapped to.

```
The setting is used to control the dpkg options --force-confold and --force-confnew and are applied when installing apt packages.
```

Comment (PradeepKiruvale) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:739]: updated

**Comment by PradeepKiruvale** [comment]
> Can you please add the source files how you generated the sampledeb package (e.g. nfpm config file and the script used to build them) to the test folder. Just place them next to where the .deb files are located.
> 
> This will make it easier to make adjustments to the package in the future.

Added those nfpm files

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot]: Yes, it works. But now I fixed it, as suggested above.

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot]: removed

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot]: Moved these steps to `tear down`

Comment (reubenmiller) [reply tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot]: Though I think you should use the Test Teardown and not the Suite Teardown as you want to make the device state is in the expected state before running the second test.

**Comment by reubenmiller** [comment]
> > Can you please add the source files how you generated the sampledeb package (e.g. nfpm config file and the script used to build them) to the test folder. Just place them next to where the .deb files are located.
> > This will make it easier to make adjustments to the package in the future.
> 
> Added those nfpm files

Can you also add a shell script to build those two packages, as it is self-documenting on how to reproduce those deb files etc.

**Comment by PradeepKiruvale** [comment]
> > > Can you please add the source files how you generated the sampledeb package (e.g. nfpm config file and the script used to build them) to the test folder. Just place them next to where the .deb files are located.
> > > This will make it easier to make adjustments to the package in the future.
> > 
> > 
> > Added those nfpm files
> 
> Can you also add a shell script to build those two packages, as it is self-documenting on how to reproduce those deb files etc.

added
#

File: tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/install_overwrite_config_files.robot]: right, I have fixed that

File: tests/RobotFramework/tests/plugin_apt/nfpm-version1.yaml
Comment (reubenmiller) [inline tests/RobotFramework/tests/plugin_apt/nfpm-version1.yaml:13]: The nfpm file is referencing files that are not present in the project.

You don't need to reflect the target destination in the `src` field. You can just add configuration files for each version, e.g. `sampledeb_1.cfg` and `sampledeb_2.cfg`.

File: tests/RobotFramework/tests/plugin_apt/nfpm-version2.yaml
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/nfpm-version2.yaml:13]: added `sampledeb_2.cfg` file

File: tests/RobotFramework/tests/plugin_apt/nfpm-version1.yaml
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/nfpm-version1.yaml:13]: added `sampledeb_1.cfg` file

File: tests/RobotFramework/tests/plugin_apt/nfpm-version2.yaml
Comment (reubenmiller) [reply tests/RobotFramework/tests/plugin_apt/nfpm-version2.yaml]: Yes, though unfortunately the files you added don't match your nfpm config which resulted in the errors when trying to build the debian packages via the build.sh script.

I've fixed the config (and made some formatting improvements in the robot file)

File: tests/RobotFramework/tests/plugin_apt/nfpm-version1.yaml
Comment (reubenmiller) [reply tests/RobotFramework/tests/plugin_apt/nfpm-version1.yaml]: Yes, though unfortunately the files you added don't match your nfpm config which resulted in the errors when trying to build the debian packages via the build.sh script.

I've fixed the config (and made some formatting improvements in the robot file)

File: tests/RobotFramework/tests/plugin_apt/nfpm-version2.yaml
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/plugin_apt/nfpm-version2.yaml]: :( I had made a change to the code, but forgot to push it.

**Review by reubenmiller — APPROVED** [review]
Approved, though please wait for a Rust review before merging

I checked the following:
* system test and dependencies
* tedge config description and example

**Review by didier-wenzek — APPROVED** [review]
Approved with some suggestions

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:755]: fixed

**Pull request merged**
